src/main/scala/firrtl/options/StageAnnotations.scala

```diff
@@ -14,6 +17,63 @@ sealed trait StageOption { this: Annotation => }
   */
 trait Unserializable { this: Annotation => }
 
+/** Mix-in that lets an [[firrtl.annotations.Annotation Annotation]] serialize itself to a file separate from the output
+  * annotation file.
+  *
+  * This can be used as a mechanism to serialize an [[firrtl.options.Unserializable Unserializable]] annotation or to
+  * write ancillary collateral used by downstream tooling, e.g., a TCL script or an FPGA constraints file. Any
+  * annotations using this mix-in will be serialized by the [[firrtl.options.phases.WriteOutputAnnotations
+  * WriteOutputAnnotations]] phase. This is one of the last phases common to all [[firrtl.options.Stage Stages]] and
+  * should not have to be called/included manually.
+  *
+  * Note: from the perspective of transforms generating annotations that mix-in this trait, the serialized files are not
+  * expected to be available to downstream transforms. Communication of information between transforms must occur
+  * through the annotations that will eventually be serialized to files.
+  */
+trait CustomFileEmission { this: Annotation =>
+
+  /** Output filename where serialized content will be written
+    *
+    * The full annotation sequence is a parameter to allow for the location where this annotation will be serialized to
+    * be a function of other annotations, e.g., if the location where information is written is controlled by a separate
+    * file location annotation.
+    *
+    * @param annotations the annotation sequence at the time of emission
+    */
+  protected def baseFileName(annotations: AnnotationSeq): String
+
+  /** Optional suffix of the output file */
+  protected def suffix: Option[String]
+
+  /** A method that can convert this annotation to bytes that will be written to a file.
+    *
+    * If you only need to serialize a string, you can use the `getBytes` method:
+    * {{{
+    *  def toBytes: Option[Iterable[Byte]] = Some(myString.getBytes)
+    * }}}
+    */
+  def toBytes: Option[Iterable[Byte]]
+
+  /** Optionally, a sequence of annotations that will replace this annotation in the output annotation file.
+    *
+    * A non-empty implementation of this method is a mechanism for telling a downstream [[firrtl.options.Stage Stage]]
+    * how to deserialize the information that was serialized to a separate file. For example, if a FIRRTL circuit is
+    * serialized to a separate file, this method could include an input file annotation that a later stage can use to
+    * read the serialized FIRRTL circuit back in.
+    */
+  def replacements(file: File): AnnotationSeq = Seq.empty
```

Comment:
Is the API that the file has been written? I think there is potential here for some awkward interactions.

* Is it considered acceptable practice for implementers of this method to read the file?
* As implemented, this gives the canonical file. Canonical files are absolute so if people use this argument they lose the ability to make the information relocatable

I'm really not certain what the API here should be, but I'm thinking it should be possible to get a relative path, and maybe just have it be a `String` instead of a `File`. Not sure though.

Reply:
This is very tightly coupled to the implementation of `WriteOutputAnnotations`.

> Is the API that the file has been written? I think there is potential here for some awkward interactions.

Yes. At this point, the custom file has been written. The output annotation file has not. (We can change this to not write the file yet, if needed---see next paragraph.)

> - Is it considered acceptable practice for implementers of this method to read the file?

No... it's not generally acceptable. Thought, it is allowable. Two thoughts: (1) the file could be rewritten after this so that the transform doesn't have access to the file contents and (2) we can use a custom security manager that disallows file writes when we call `replacements` (this is extremely draconian, though). I think probably (1) makes sense?

> - As implemented, this gives the canonical file. Canonical files are absolute so if people use this argument they lose the ability to make the information relocatable

Yeah, this won't work if you elaborate on one machine, then run FIRRTL compilation on another. However, I think this is actually an issue of `StageOptions.getBuildFileName` which is returning a canonical path as opposed to just a path. Let me see if changing that produces the correct behavior.

Reply:
The change to relative paths is actually one line... Fixed here: https://github.com/freechipsproject/firrtl/pull/1277/commits/5637e3f8c5221312e0cd9119ea75528aad390e87.

Reply:
I'm a little worried that changing that utility might have downstream affects on dependent projects. In particular I think @chick should be aware since (IIRC) he wrote the original code.

I suspect that some of the original intent of the canonicalization there was to do normalization (eg. turn `foo/../bar` into just `bar`. We can do that without constructing an absolute path by doing: `.toPath.normalize.toFile`.

Reply:
Good point. I've been locally building the world with this, but it may have unexpected consequences. 

Nice idea about `.toPath.normalize.toFile`. I've added a co-authored commit with this change.